src/filter.rs

```diff
@@ -0,0 +1,300 @@
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error};
+use std::ffi::{CString, CStr};
+use std::path::PathBuf;
+use raw::{git_filter, git_writestream};
+
+
+#[repr(C)]
+struct Filter<T> {
+    filter: git_filter,
+    instance: T,
+}
+
+#[repr(C)]
+struct Stream<T> {
+    parent: git_writestream,
+    handler: T,
+}
+
+/// Wrapper of the raw git_writestream object, representing the input
+/// of next filter of the chain
+pub struct RawStream {
+    stream: *mut git_writestream,
+}
+
+/// A trait representing operations to handle filter data stream.
+///
+/// These callbacks are typically called by libgit2.
+pub trait WriteStream {
+    /// New stream data, Implementations should process the buffer data
+    /// and then write the output to the next stream if needed.
+    fn write(&mut self, buf: &[u8])
+        -> Result<(), Error>;
+
+    /// Close stream. Implementations should do finalizing work and then
+    /// close the next stream.
+    fn close(&mut self) -> Result<(), Error>;
+}
+
+/// Information to construct a new stream filter.
+pub struct FilterSource {
+    /// Path to the source file to be filter.
+    pub path: PathBuf,
+    /// A sumgde filter if true, otherwise a clean filter.
+    pub to_worktree: bool,
+}
+
+/// A trait representing a customized filter to be registered into libgit2
+/// filter chain.
+///
+/// Modeled after git2/sys/filter.h in a streaming manner.
+pub trait StreamFilter {
+    /// Customized WriteStream implementation.
+    type S: WriteStream;
```

Comment:
Instead of having an associated type this may want to just return `Box<WriteStream>` below to avoid extra monomorphization costs